src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,320 @@
+                return curlioerr.CURLIOE_FAILRESTART; // Force curl_easy_perform to fail
+            }
+        }
+
+        private static ulong ExecuteCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context,
+            Func<IntPtr, ulong, RequestCompletionSource, ulong> callback)
+        {
+            size *= nitems;
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                return (size == 0) ? 0 : callback(buffer, size, state);
+            }
+            catch (Exception ex)
+            {
+                HandleAsyncException(state, ex);
+                return (size > 0) ? size - 1 : 1; // Force curl_easy_perform to fail
+            }
+        }
+
+        private static bool TryParseStatusLine(HttpResponseMessage response, string responseHeader)
+        {
+            if (!responseHeader.StartsWith(s_httpPrefix, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            // Check if line begins with HTTP/1.1 or HTTP/1.0
+            int prefixLength = s_httpPrefix.Length;
+            if ((responseHeader[prefixLength] == '1') && (responseHeader[prefixLength + 1] == '.'))
+            {
+                if (responseHeader[prefixLength + 2] == '1')
+                {
+                    response.Version = new Version(1, 1);
+                }
+                else if (responseHeader[prefixLength + 2] == '0')
+                {
+                    response.Version = new Version(1, 0);
+                }
+            }
+
+            // Parse first 3 characters after a space as status code
+            // TODO: Parsing errors are treated as fatal. Find right behaviour
+            int statusCodeLength = 3;
+            int codeIndex = responseHeader.Substring(prefixLength).IndexOf(' ') + 1;
+            string strStatusCode = responseHeader.Substring(prefixLength + codeIndex, statusCodeLength);
+            response.StatusCode = (HttpStatusCode)(int.Parse(strStatusCode));
+
+            // The remaining string after the space is the reason phrase
+            response.ReasonPhrase = responseHeader.Substring(prefixLength + codeIndex + statusCodeLength + 1);
+
+            return true;
+        }
+
+        private static int CurlSocketCallback(
+            IntPtr handle,
+            curl_socket_t socketFd,
+            int socketAction,
+            IntPtr context,
+            IntPtr sockPtr)
+        {
+            int retVal = CURLMcode.CURLM_OK;
+
+            if (CurlPoll.CURL_POLL_REMOVE == socketAction)
+            {
+                return retVal;
+            }
+
+            bool isAdd = false;
+            if (IntPtr.Zero == sockPtr)
+            {
+                int result = Interop.libcurl.curl_easy_getinfo(handle, CURLINFO.CURLINFO_PRIVATE, out sockPtr);
```

Comment:
Yes.. in fact, they have to be used together. Since a multi handle is used for multiple requests, you need an easy handle to set per-request settings.